### PR TITLE
Feature/contacts repo

### PR DESCRIPTION
**Pull request by Valink16**
This pull request introduces the `ContactsRepositoryLocal`, a local data source for managing contacts using `DataStore`. It implements the `ContactsRepository` interface, which decouples the data layer from the UI and simplifies integration. The repository provides a full set of CRUD operations for `Contact` objects, ensuring data is persisted securely on the device.

### Key Features:
- **CRUD Operations**: Implements `addContact`, `getContact`, `editContact`, `deleteContact`, and `getAllContacts`.
- **Secure Storage**: Contact data is encrypted before being saved to `DataStore` to ensure user privacy.
- **Robust Error Handling**: The repository gracefully handles various failure scenarios, such as attempting to access non-existent contacts, adding duplicates, or reading corrupted data.

### Testing:
- Added a comprehensive suite of unit tests (`ContactsRepositoryLocalTests`) using Robolectric.
- Tests cover all public methods, including success paths, failure cases, and edge cases like data corruption and mismatched IDs during updates.

This implementation provides a reliable and secure foundation for contact management within the application.

### CODE REVIEW

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        val ciphered = CryptoUtils.encrypt(gson.toJson(contact))
+        it[keyFor(contact)] = ciphered
+        r = Result.success(Unit)
+      }
+    } catch (e: Exception) {
+      Log.e("ContactStorage", "Error saving contact ${contact.id}", e)
+      r = Result.failure(StorageException.DataStoreError(e))
+    }
+
+    return r
```
Comment (hohoho2002) [inline app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: This function used a mutable variable `r` declared outside the `dataStore.edit {}` block and updated it inside the block. Since `edit {}` is a suspending function that can run its lambda on a different coroutine context/dispatcher, this creates a cross-thread mutation. I think this is unsafe because there is no synchronization, which can lead to stale or inconsistent results. I suggest removing the external mutable variable and instead wrapping the `edit` block in `runCatching { ... }`. This approach is thread-safe because all state is local to the coroutine and no cross-thread mutation occurs
```suggestion
    val result = runCatching {
        dataStore.edit { prefs ->
            if (prefs.contains(keyFor(contact))) {
                throw StorageException.DataStoreError(Exception("Contact ${contact.id} already exists"))
            }
            val ciphered = withContext(Dispatchers.IO) { CryptoUtils.encrypt(gson.toJson(contact)) }
            prefs[keyFor(contact)] = ciphered
        }
    }.mapFailure {
        if (it is StorageException) it else StorageException.DataStoreError(it)
    }
    return result
```

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        } catch (e: Exception) {
+          Log.e("ContactStorage", "Failed to parse JSON for contact $contactID", e)
+          return Result.failure(StorageException.DeserializationError(e))
+        }
+
+    return Result.success(contact)
+  }
+
+  override suspend fun editContact(contactID: String, newContact: Contact): Result<Unit> {
+    var r: Result<Unit> = Result.success(Unit)
```
Comment (hohoho2002) [inline app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: The same as `addContact()`

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            return@edit
+          }
+    }
+
+    return r
+  }
+
+  override suspend fun deleteContact(contactID: String): Result<Unit> {
+    val key = stringPreferencesKey(contactID)
+    var r = Result.success(Unit)
```
Comment (hohoho2002) [inline app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: The same as `addContact`

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    return r
+  }
+
+  override suspend fun getAllContacts(): Result<List<Contact>> {
+    val prefs = dataStore.data.first()
+    val contacts = mutableListOf<Contact>()
+    for ((_, value) in prefs.asMap()) {
+      try {
+        val decrypted = CryptoUtils.decrypt(value as String)
```
Comment (hohoho2002) [inline app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: This assumes all stored preferences are your encrypted contacts — if something else is stored, this will crash. We can check type like this
```suggestion
        val stringValue = value as? String ?: return@mapNotNull null
```

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        val contact = gson.fromJson(decrypted, Contact::class.java)
+        contacts.add(contact)
+      } catch (e: Exception) {
+        throw StorageException.DataStoreError(e)
+      }
+    }
+    return Result.success(contacts)
```
Comment (hohoho2002) [inline app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: I suggest this because it always return a `Result`. In current version, `StorageException.DataStoreError(e)` can be thrown and `Result` won't be return
```suggestion
    val prefs = dataStore.data.first()
    prefs.asMap().mapNotNull { (_, value) ->
        val stringValue = value as? String ?: return@mapNotNull null
        val decrypted = CryptoUtils.decrypt(stringValue)
        gson.fromJson(decrypted, Contact::class.java)
    }
}.mapFailure {
    StorageException.DataStoreError(it)
```

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            try {
+              val ciphered = CryptoUtils.encrypt(gson.toJson(newContact))
+              ciphered
+            } catch (e: Exception) {
+              Log.e("ContactStorage", "Error encrypting contact", e)
+              throw StorageException.EncryptionError(e)
+            }
+      }
+    }
+  }
```
Comment (hohoho2002) [inline app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: Cleaner now !

File: app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            dataStore.data.map { it[key] }.first()
+          } catch (e: ClassCastException) {
+            Log.e("ContactStorage", "Failed to load ciphered JSON for $contactID", e)
+            throw StorageException.DataStoreError(e)
+          }
+
+      if (ciphered == null) {
+        Log.e("ContactStorage", "Contact $contactID not found")
+        throw StorageException.DataStoreError(Exception("Contact $contactID not found"))
+      }
```
[the remark this answers is not in the document]
Comment (Valink16) [reply app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt]: The try catches are there because we want to rethrow StorageExceptions.

**Review by hohoho2002 — APPROVED** [review]
